In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [2]:
# Function to normalize the rating matrix by centering each row
def normalize_ratings_by_row_mean(matrix):
    row_means = matrix.mean(axis=1, skipna=True)  # Calculate row means excluding NaN values
    normalized_matrix = matrix.sub(row_means, axis=0)
    return normalized_matrix

In [3]:
# Read the R-matrix
Rmat = pd.read_csv('Rmat.csv', index_col=0)

# Normalize R-matrix
Rmat = normalize_ratings_by_row_mean(Rmat)

Rmat

m1  m2  m3        m4  m5        m6  m7  m8  m9  m10  ...  m3943  \
u1     0.811321 NaN NaN       NaN NaN       NaN NaN NaN NaN  NaN  ...    NaN   
u2          NaN NaN NaN       NaN NaN       NaN NaN NaN NaN  NaN  ...    NaN   
u3          NaN NaN NaN       NaN NaN       NaN NaN NaN NaN  NaN  ...    NaN   
u4          NaN NaN NaN       NaN NaN       NaN NaN NaN NaN  NaN  ...    NaN   
u5          NaN NaN NaN       NaN NaN -1.146465 NaN NaN NaN  NaN  ...    NaN   
...         ...  ..  ..       ...  ..       ...  ..  ..  ..  ...  ...    ...   
u6036       NaN NaN NaN -1.302928 NaN -0.302928 NaN NaN NaN  NaN  ...    NaN   
u6037       NaN NaN NaN       NaN NaN       NaN NaN NaN NaN  NaN  ...    NaN   
u6038       NaN NaN NaN       NaN NaN       NaN NaN NaN NaN  NaN  ...    NaN   
u6039       NaN NaN NaN       NaN NaN       NaN NaN NaN NaN  NaN  ...    NaN   
u6040 -0.577713 NaN NaN       NaN NaN       NaN NaN NaN NaN  NaN  ...    NaN   

       m3944  m3945  m3946  m3947  m3948  m3949  m3950  m3951  m3952  
u1       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN  
u2       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN  
u3       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN  
u4       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN  
u5       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN  
...      ...    ...    ...    ...    ...    ...    ...    ...    ...  
u6036    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN  
u6037    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN  
u6038    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN  
u6039    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN  
u6040    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN  

[6040 rows x 3706 columns]

In [4]:
import time
def compute_similarity(matrix):
    matrix_values = matrix.values
    num_movies = len(matrix.columns)
    similarity_matrix = np.ones((num_movies, num_movies))*np.nan

    for i in range(num_movies):
        start_time = time.time()
        for j in range(i+1, num_movies):
            common_users = np.where(~np.isnan(matrix_values[:, i]) & ~np.isnan(matrix_values[:, j]))[0]
            if len(common_users) > 2:
                ratings_i = matrix_values[common_users, i]
                ratings_j = matrix_values[common_users, j]

                numerator = np.sum(ratings_i * ratings_j)
                denominator = np.sqrt(np.sum(ratings_i**2)) * np.sqrt(np.sum(ratings_j**2))

                if denominator != 0:
                    similarity = 0.5 + 0.5 * (numerator / denominator)
                    similarity_matrix[i, j] = similarity
                    similarity_matrix[j, i] = similarity
        if (i+1)%50==0 or i+1==num_movies:
            finish_time = time.time()
            run_time = finish_time-start_time
            print(f'Progress: {i+1}/{num_movies} (Run Time: {round(run_time, 2)})')
    return similarity_matrix

In [5]:
Smat = compute_similarity(Rmat)

Progress: 50/3706 (Run Time: 0.09)
Progress: 100/3706 (Run Time: 0.06)
Progress: 150/3706 (Run Time: 0.07)
Progress: 200/3706 (Run Time: 0.07)
Progress: 250/3706 (Run Time: 0.07)
Progress: 300/3706 (Run Time: 0.07)
Progress: 350/3706 (Run Time: 0.05)
Progress: 400/3706 (Run Time: 0.07)
Progress: 450/3706 (Run Time: 0.06)
Progress: 500/3706 (Run Time: 0.06)
Progress: 550/3706 (Run Time: 0.04)
Progress: 600/3706 (Run Time: 0.04)
Progress: 650/3706 (Run Time: 0.06)
Progress: 700/3706 (Run Time: 0.07)
Progress: 750/3706 (Run Time: 0.02)
Progress: 800/3706 (Run Time: 0.06)
Progress: 850/3706 (Run Time: 0.06)
Progress: 900/3706 (Run Time: 0.04)
Progress: 950/3706 (Run Time: 0.05)
Progress: 1000/3706 (Run Time: 0.04)
Progress: 1050/3706 (Run Time: 0.05)
Progress: 1100/3706 (Run Time: 0.05)
Progress: 1150/3706 (Run Time: 0.04)
Progress: 1200/3706 (Run Time: 0.06)
Progress: 1250/3706 (Run Time: 0.05)
Progress: 1300/3706 (Run Time: 0.04)
Progress: 1350/3706 (Run Time: 0.04)
Progress: 1400/3706 (

In [6]:
Smat_df = pd.DataFrame(Smat, index=Rmat.columns, columns=Rmat.columns)
Smat_df

m1        m2        m3        m4        m5        m6        m7  \
m1          NaN  0.442636  0.410766  0.404020  0.362457  0.579168  0.475257   
m2     0.442636       NaN  0.546743  0.460495  0.648002  0.412014  0.571488   
m3     0.410766  0.546743       NaN  0.640756  0.677383  0.437453  0.547960   
m4     0.404020  0.460495  0.640756       NaN  0.710145  0.318943  0.513088   
m5     0.362457  0.648002  0.677383  0.710145       NaN  0.388942  0.587943   
...         ...       ...       ...       ...       ...       ...       ...   
m3948  0.548122  0.553910  0.528168  0.457428  0.590165  0.534482  0.477317   
m3949  0.702972  0.323460  0.284015  0.168765  0.239982  0.615329  0.355397   
m3950  0.555231  0.316563  0.515971  0.621891  0.517104  0.419051  0.649063   
m3951  0.703109  0.223420  0.490683  0.660954       NaN  0.380572  0.676682   
m3952  0.600869  0.421348  0.427316  0.449424  0.347755  0.499672  0.381187   

             m8        m9       m10  ...     m3943     m3944     m3945  \
m1     0.303379  0.188966  0.512106  ...  0.381834  0.089488  0.227803   
m2     0.488686  0.689619  0.541504  ...  0.398421       NaN  0.441406   
m3     0.669407  0.587030  0.568835  ...  0.439756  0.458236  0.385446   
m4     0.638414  0.494078  0.369146  ...  0.764084       NaN       NaN   
m5     0.608642  0.758438  0.503592  ...  0.131195  0.509458  0.601690   
...         ...       ...       ...  ...       ...       ...       ...   
m3948  0.323261  0.441275  0.574412  ...  0.465711  0.536067  0.448532   
m3949  0.110612  0.213767  0.439866  ...  0.563115  0.301218  0.252580   
m3950  0.796771       NaN  0.589969  ...  0.480608  0.233652  0.020660   
m3951  0.131729       NaN  0.022272  ...  0.426416  0.387007  0.173395   
m3952  0.258232  0.350740  0.565138  ...  0.605438  0.589734  0.552011   

          m3946     m3947     m3948     m3949     m3950     m3951     m3952  
m1     0.270004  0.649773  0.548122  0.702972  0.555231  0.703109  0.600869  
m2     0.685420  0.261454  0.553910  0.323460  0.316563  0.223420  0.421348  
m3     0.703825  0.503654  0.528168  0.284015  0.515971  0.490683  0.427316  
m4     0.907509  0.348647  0.457428  0.168765  0.621891  0.660954  0.449424  
m5     0.855912  0.644852  0.590165  0.239982  0.517104       NaN  0.347755  
...         ...       ...       ...       ...       ...       ...       ...  
m3948  0.515220  0.369859       NaN  0.491852  0.425333  0.307334  0.506530  
m3949  0.295247  0.657607  0.491852       NaN  0.594070  0.656516  0.566232  
m3950  0.502494  0.532813  0.425333  0.594070       NaN  0.662502  0.573728  
m3951  0.398000       NaN  0.307334  0.656516  0.662502       NaN  0.690165  
m3952  0.373659  0.321705  0.506530  0.566232  0.573728  0.690165       NaN  

[3706 rows x 3706 columns]

In [7]:
Smat_df.loc[["m1", "m10", "m100", "m1510", 'm260', 'm3212'], ["m1", "m10", "m100", "m1510", 'm260', 'm3212']]

m1       m10      m100  m1510      m260  m3212
m1          NaN  0.512106  0.392000    NaN  0.741148    NaN
m10    0.512106       NaN  0.547458    NaN  0.534334    NaN
m100   0.392000  0.547458       NaN    NaN  0.329694    NaN
m1510       NaN       NaN       NaN    NaN       NaN    NaN
m260   0.741148  0.534334  0.329694    NaN       NaN    NaN
m3212       NaN       NaN       NaN    NaN       NaN    NaN

In [8]:
Smat_df.to_csv('Smat.csv')

Compute the Cosine similarity among the 3,706 movies. For movies $i$ and $j$, let $I_{ij}$ denote the set of users who rated both movies $i$ and $j$. We decide to ignore similarities computed based on less than three user ratings. Thus, define the similarity between movie $i$ and $j$ as follows, when the cardinality of $I_{ij}$ is bigger than two,

$$
S_{ij} = \frac{1}{2} + \frac{1}{2}\frac{\sum_{l \in I_{ij}}R_{li}R_{lj}}{\sqrt{\sum_{l \in I_{ij}}R^2_{li}} \sqrt{\sum_{l \in I_{ij}}R^2_{lj}}}
$$

This transformation (1 + cos)/2 ensures that similarity measures are between 0 and 1. NA values may occur when 1) the set $I_{ij}$ has a cardinality less than or equal to two (i.e., this pair of movies have been rated by only zero, one, or two users) or 2) one of the denominators is zero.